### DISCUSSION TITLE
Debug compile and install time

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#1988] [Q&A]
Hello @ajwfrost , I think we adressed this in the past as well, but I don't think we got a propper workflow here.
When compiling AAB for debug and installing them it takes 2 minutes and a bit compile and 2 minutes to install on device.
Is there a workflow we could apply to speed this up?
You mentioned something about running the project through Android Studio, is this documented somewhere and we missed it?

Thank you!

**Comment by ajwfrost** [comment]
Hi

When creating an AAB file, there's an intermediate step because we basically have to create a project structure that can be used by the Android Gradle plug-in (which is fairly quick to do), and then we invoke Gradle to build the .aab file (which can take a bit of time). Then you have to do the installation, and for an AAB file to install we first have to generate an APK set based on what the device supports, and then download the relevant APKs to the device. So that also takes a bit of time - I'd actually suggest apk-debug is better when debugging as it skips the APKS step...

But quicker than that would be to stop at the intermediate step (either you can just uncomment the adt.cfg configuration item `KeepAndroidStudioOutput=true` which means it will leave the project structure in place, or use a target type of `android-studio-output`). Then open that in Android Studio and you can then do deployment/debugging from there. If you update your SWFs you can copy them manually into the Android Studio assets location and re-build/deploy again.. (although I'm not sure it always works out that these files have changed so sometimes you need to force it to re-package e.g. by changing a version number).

I would be interested to know more about some of the common processes that people go through when building/testing/debugging on devices, as we may be able to streamline things a little better. So e.g. if I'm using FlashDevelop with a project and do a build (F8) and then "Start Remote Session", it would be quite nice if I could have a single set-up of the APK installation to start with but then for it to just know about the updates that were built, and for it to automatically re-start the app and engage the debugger when the remote session is requested..   - but is that the sort of workflow that you'd also use and would find useful?

thanks

**Comment by raresn** [comment]
Thank you for taking the time to answer!
Will definitely try to create a bat file that copies the SWF to the right place then deploy from Android Studio.

We use Flash Builder ( sry about this :) - we tried to migrate to intelliJ but we got stuck on the debug/export process and just gave up.... if we'll find some solid tutorials on how to migrate stuff, we will definitely try again!)

We debug in 2 ways:
1) preview in simulator - press f12, and we will preview the app in the profile we set from the debug settings
2) make the SWF from Builder, then call a bat file that packages the APK/AAB + installs it on phone then we either:
- start FDB directly, then start the app on the phone
- start Adobe Animate, new project, write a stop(); in the actionscript and hit "start remote debugging" - then start the app on the phone.

**Reply by raresn** [reply]
@ajwfrost - tried the android-studio-debug
Compiled, deploys on phone.
Then i changed the swf inside the "app\src\main\assets\Agenda.swf " folder but it does not update the app on phone. Keeps the old swf...
I tried changing the version inside the manifest, did a clean, rebuild, none of these work.

Also - what is the difference between android-studio and android-studio-debug? anywhere we can read about this?

**Reply by ajwfrost** [reply]
Hmmm, am not sure then how you can force Android Studio to realise that the file has changed. Maybe drop a completely new file into the same folder? or do a "reload from disk"?

The `-debug` directives on the end of the various Android package commands mean that we use a build of the runtime classes with additional outputs and with setting such that the AIR runtime is passed the 'debug' flag during start-up. So it can be used to try to connect to a remote debugger, and it will enable the outputs of trace() to logcat and so on. We also updated the bootstrap code for debug builds so that it looks to see if we've got our own runtime on the device and if so, it loads that instead (well, this worked up to android 12..) which is very handy for debugging!

Thanks for the earlier info .. I'm still amazed by how many people use Flash Builder, and wondering why no one has yet created a tool that fully converts a Flash Builder project (with all dependencies and settings) into other formats such as for IntelliJ, VS Code or Moonshine...
:-)

**Reply by raresn** [reply]
@ajwfrost  - Reload from disk didn't work, adding a new file in that folder did nothing either.

I am still amazed we use it as well - it is an obsolete tool - I will delete it from my SSD right now if i had the proper procedures to compile/debug in IntelliJ :)
If you guys decide to build such a tool - please let me know because we would love to test it :)

**Reply by raresn** [reply]
@ajwfrost how would we debug in intelij? would we be able to connect the phone via usb and put brakepoints in our code and debug that way?

**Reply by marchbold** [reply]
@raresn You definitely can use direct device debugging and breakpoints etc in intellij. The process is you setup a module with your application build settings, then you setup a run configuration to deploy it to a device and run in debug mode.

**Reply by ydkgame** [reply]
![image](https://user-images.githubusercontent.com/3896827/178094336-b0b2db23-051b-4018-b3e1-8e730b569a6d.png)
After replacing SWF, change it again to get the normal update

**Comment by Adrian-S** [comment]
@ajwfrost - Since I work with @raresn I have the same pattern. Flash Builder still works great but has some downsides and I think you will hear more about it since latest SDKs can't be overlayed in FB. But we've been creating SWFs with older SDKs since ever and they compile just fine with latest SDKs. Other issues are the ANEs and testing on mobile. But most of our work is not device specific, therefore we still use AS3 and haven't moved to other platforms.

**The problem:**
When it comes to compiling with ANEs and permissions and manifests, I have to confess, that gives me the biggest headaches. I don't know enough about android manifests or iPhone additions, and there are a lot of options we never heard of and probably never will. The APM tool from @distriqt is nice, but still a bit complicated to use. It’s certainly faster than manual install, but I still have to keep a separated XML with my other manual edits, or ANE files. And most edits in these files are based on hopes and prayers. We will probably buy intelyJ and give it another go. PhpStorm is nice, to bad I can't use that since I already have it bought.

**Idea 1:**
What would help us at this time, is a GUI project manager / compiler. I honestly want to forget there are XML files. This will act like a knowledge base for all those unknown manifest options. Like how `android:hardwareAccelerated="true"` is good for scrolling in WebViews. ANEs should request all manifest additions and give configuration options if necessary. APM does this, AneLAB did it in a GUI. This could also work as a ANE marketplace.

I don’t know if any other IDE does this or how the compile flow is done. But we have always used a dedicated BAT file for store releasing, and suspect we will continue to do so, because it's easier than changing certificates from development to production.

**Idea 2:**
Another suggestion for faster "compile" times would be just to skip compile and send the SWF to the app. Or just replace the SWF in the existing APK files. A checksum for all resources should/could be made before switching the SWF file. I think in 99% of cases nothing would change. Sending the SWF file in an already installed and running app, causing a quick relaunch, would be almost instant.

PS. FB has some good keyboard shortcuts, that I got accustomed to. If I could combine it with multiline edit from Sublime Text, I would not even start to consider other IDEs.

**Reply by ajwfrost** [reply]
Hi

Thanks for the suggestions. I would hope that the first idea here is going to be a second phase for `apm` - we were also wondering whether there's a value to enhancing some of the compilation tools so that you can actually see what's going on under the hood i.e. how are these actually being invoked, what options are there, etc. So that you can do something a bit like "record" what is invoked by IDE1 (such as Flash Builder), and then "replay" to just do that again outside of an IDE (or to export it for use by IDE2 such as IntelliJ, VSCode, Moonshine etc).

For the second idea, yes this is exactly what we've been looking at for enhancing the on-device debugging experience and reducing round-trip times... if you've got an app already packaged and installed on the handset, then it should be possible for there to be something monitoring your filesystem on the PC, and if the SWF files are changed, for these to then be passed over to the device and the app restarted. Definitely on our plans!

I still sometimes wonder whether we should have picked up Flash Builder from Adobe as well .. but it was complex due to the purchasing/licensing issues (plus I don't really like it and would be happier if people dropped anything Eclipse-based as soon as possible!)

thanks

**Reply by marchbold** [reply]
@Adrian-S Would love to hear what you find complicated about apm. We have attempted to make it as simple as possible. There should be very few circumstances that you need to manually edit xml with apm now beyond advanced configuration settings. 

We definitely have a gui tool for apm on the horizon but really want to flesh out the core functionality with the command line tool first.

**Reply by marchbold** [reply]
PS shortcuts in intellij I have found to be much better than FB ever was, definitely suggest checking them out :)

**Reply by marchbold** [reply]
@ajwfrost Personally I think FB is worth leaving in the past, the eclipse ide it's based on has been superseded by a lot of better tools now. If you are going to do something I'd highly recommend either approaching JetBrains about contributing to their plugin or further developing the VSCode plugin @joshtynjala has started (probably the later is the better business proposition).

**Reply by Adrian-S** [reply]
@marchbold - apm is just command line, and for FCM icons need to be edited manually. Res folder needs to be included in xml and I don't know how to do this with apm. I don't know what happens with other ANEs. Fortunately for us all extensions are now from distriqt and we don't have to worry about other ANEs for now. Also noticed *.default.swc is no longer in any ANE, we had that, from an older version, included in FB libs and permitted FB to work better. And by better we didn’t have to hide extension from XML to compile for Desktop.

@ajwfrost - I'm glad these are in plan for the future. Monitoring for SWF changes should be an option like "AutoRun on selected Device(s)". I was also thinking of fetching the SWF remotely from desktop. Since 'fdb' gets a connection to the app it could also provide a file upload just before starting the app.

One thing you should have grabbed from Adobe, is Adobe Scout. I don't know if there is something better for advanced telemetry, probably intellij has something, we are yet to discover it. Other IDEs might not have telemetry. FB even though obsolete, it still works 10 years later after its latest release. For some unknown reason I hate eclipse too, but accepted FB. And the learning curve for any IDE is extremely steep, but we are forcing ourselves to switch to intellij, and hitting new walls.

**Reply by ajwfrost** [reply]
> One thing you should have grabbed from Adobe, is Adobe Scout

Done :-) we've published this for Android already, because we needed to put in changes to cope with the newer Android security (for the Scout mobile config app to be able to share the settings with the AIR app), we've not yet looked at getting it building on other platforms yet (well, we had a brief look on macOS and gave up quickly, it uses quite old code/frameworks that don't work with recent Xcode/Clang versions..)

I had wondered how best to do this - whether to modify an existing tool like Scout, or to enhance FDB (but this then could need all the other IDEs to update themselves to take advantage of it..), or to create a new tool for this..  -> current thinking though is that we could provide this capability via a SWC library to just be linked in (e.g. behind a debug/release config flag) which would then reach out to FDB and that could hopefully provide the details automatically of what files to monitor for changes.

**Reply by Adrian-S** [reply]
@ajwfrost Honestly forgot there is Scout for mobile, I was thinking about the desktop app. I usually run most performance tests in air simulator, and whatever I can optimise on desktop, is beneficial for mobile as well. The only thing here is that debug mode is somehow still limited to 30fps, while release can run at 60fps. And Scout on desktop seams uncertain of what's going on. (a lot of "overhead", whatever that is). Still it's the only software, that I know, that can pinpoint what functions are making the app slow.

**Reply by marchbold** [reply]
@Adrian-S Eventually we do plan to include more of the app xml features but currently we have only implemented the core aspects related to libs/anes. 

Also, regarding swc's, we have cleaned up our extensions over the years and no longer include the default swc with the ane's (to minimise size etc). You should find that intellij doesn't need the swc like flashbuilder did, you should be able to use the ane directly. If you really do need the swc, they are available in the repositories.
